### PR TITLE
New payment pingpong scaffold

### PR DESCRIPTION
**Pull request by vkuznecovas**
Initial scaffold of the new pingpong between the consumer and provider for the upcoming payments.

Should be backwards compatible, as I've tested conencting to an older node, as well as had this session while running in provider mode: https://testnet.mysterium.network/session/2f822bd5-c5ff-4ccb-a921-c5ee7255790a

### CODE REVIEW

File: session/pingpong/invoice_tracker.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	}
+}
+
+func calculateMaxNotReceivedExchangeMessageCount(chargeLeeway, chargePeriod time.Duration) uint64 {
+	return uint64(math.Round(float64(chargeLeeway) / float64(chargePeriod)))
+}
+
+// Start stars the invoice tracker
+func (it *InvoiceTracker) Start() error {
+	log.Debug("starting...")
```
Comment (tadaskay) [inline session/pingpong/invoice_tracker.go]: Let's use uppercase for logging messages

File: session/pingpong/invoice_tracker.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	log.Debug(it.maxNotReceivedExchangeMessages)
```
Comment (tadaskay) [inline session/pingpong/invoice_tracker.go]: This should have some context in addition to logging an `int`.

File: session/pingpong/invoice_messaging.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/mysteriumnetwork/payments/crypto"
+)
+
+// InvoiceRequest structure represents the invoice message that the provider sends to the consumer
+type InvoiceRequest struct {
+	Invoice crypto.Invoice `json:"invoice"`
+}
+
+const endpointInvoice = "session-invoice"
+const messageEndpointInvoice = communication.MessageEndpoint(endpointInvoice)
```
Comment (tadaskay) [inline session/pingpong/invoice_messaging.go]: I see where this is coming from, but let's use normalized order for naming, it makes it easier to comprehend:
- `endpointInvoice` → `invoiceEndpoint`
- `messageEndpointInvoice` → `invoiceMessageEndpoint`

File: session/pingpong/invoice_tracker.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	peerInvoiceSender               PeerInvoiceSender
+	exchangeMessageChan             chan crypto.ExchangeMessage
+	chargePeriod                    time.Duration
+	exchangeMessageWaitTimeout      time.Duration
+	notReceivedExchangeMessageCount uint64
+	maxNotReceivedExchangeMessages  uint64
+	once                            sync.Once
+}
+
+// NewInvoiceTracker creates a new instancec of invoice tracker
```
Comment (soffokl) [inline session/pingpong/invoice_tracker.go]:
```suggestion
// NewInvoiceTracker creates a new instance of invoice tracker
```

File: session/pingpong/invoice_tracker.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// ErrExchangeWaitTimeout indicates that we did not get an exchange message in time
+var ErrExchangeWaitTimeout = errors.New("did not get a new exchange message")
+
+// ErrExchangeValidationFailed indicates that there was an error with the exchange signature
+var ErrExchangeValidationFailed = errors.New("exchange validation failed")
+
+// errBoltNotFound indicates that bolt did not find a record
+var errBoltNotFound = errors.New("not found")
```
Comment (soffokl) [inline session/pingpong/invoice_tracker.go]: `errBoltNotFound` is unused (varcheck)

File: session/pingpong/exchange_message_tracker.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	if err != nil {
+		return errors.Wrap(err, "could not create exchange message")
+	}
+	log.Debug("Sending message", *msg)
+	err = emt.peerExchangeMessageSender.Send(*msg)
+	if err != nil {
+		log.Warn("Failed to send exchange message: ", err)
+	}
+
+	return nil
```
Comment (soffokl) [inline session/pingpong/exchange_message_tracker.go]: Shouldn't we return `err` here?
```suggestion
	return err
```

File: session/promise/promise_messaging.go [hunk 7]
```diff
@@ -37,6 +37,7 @@ type Message struct {
 type PaymentInfo struct {
 	LastPromise LastPromise `json:"lastPromise"`
 	FreeCredit  uint64      `json:"freeCredit"`
+	Supports    string      `json:"supported"`
```
Comment (soffokl) [inline session/promise/promise_messaging.go]: `Supports` and `supported` here, probably should be only one.

File: session/manager.go [hunk 8]
```diff
@@ -83,7 +83,10 @@ type Storage interface {
 }
 
 // BalanceTrackerFactory returns a new instance of balance tracker
-type BalanceTrackerFactory func(consumer, provider, issuer identity.Identity) (BalanceTracker, error)
+type BalanceTrackerFactory func(consumer, provider, issuer identity.Identity) (PaymentEngine, error)
+
+// PaymentEngineFactory creates a new instance of payment engine
```
Comment (anjmao) [inline session/manager.go]: Minor suggestion. End each comment sentence with a dot at the end. Godoc page generates methods documentation based on these comments.

File: session/pingpong/exchange_message_tracker.go (hunk 6 above)
Comment (vkuznecovas) [reply session/pingpong/exchange_message_tracker.go]: No, the consumer is allowed to miss a few messages therefore returning nil here. THe provider will kill the session if it exceed an allowed amount.

File: session/pingpong/exchange_message_tracker.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+				return err
+			}
+		}
+	}
+}
+
+func (emt *ExchangeMessageTracker) issueExchangeMessage(invoice crypto.Invoice) error {
+	msg, err := crypto.CreateExchangeMessage(invoice, 10, "", emt.keystore, common.HexToAddress(emt.identity.Address))
+	if err != nil {
+		return errors.Wrap(err, "could not create exchange message")
```
Comment (anjmao) [inline session/pingpong/exchange_message_tracker.go]: Just to let you know. Go 1.13 introduced errors wrapping in standard library  https://golang.org/doc/go1.13#error_wrapping

File: session/pingpong/factory.go [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/mysteriumnetwork/node/communication"
+	"github.com/mysteriumnetwork/node/core/connection"
+	"github.com/mysteriumnetwork/node/identity"
+	"github.com/mysteriumnetwork/node/session"
+	"github.com/mysteriumnetwork/payments/crypto"
+)
+
+// InvoiceFactoryCreator returns a payment engine factory
+func InvoiceFactoryCreator(dialog communication.Dialog, balanceSendPeriod, promiseTimeout time.Duration) func() (session.PaymentEngine, error) {
```
Comment (anjmao) [inline session/pingpong/factory.go]: It's a bit strange to see invoice and payments related logic in pingpong package.

File: session/pingpong/invoice_tracker.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+	// TODO: fill the fields
+	return it.peerInvoiceSender.Send(crypto.Invoice{AgreementID: 1234})
+}
+
+func (it *InvoiceTracker) receiveExchangeMessageOrTimeout() error {
+	select {
+	case pm := <-it.exchangeMessageChan:
+		if res := pm.ValidateExchangeMessage(common.HexToAddress(it.peer.Address)); !res {
+			return ErrExchangeValidationFailed
+		}
```
Comment (anjmao) [inline session/pingpong/invoice_tracker.go]: If ValidateExchangeMessage returns boolean when it's common to name variable ok so you know what this is a boolean result.

```ts
if ok := pm.ValidateExchangeMessage(common.HexToAddress(it.peer.Address)); !ok {
    return ErrExchangeValidationFailed
}
```

File: session/manager.go (hunk 8 above)
Comment (vkuznecovas) [reply session/manager.go]: makes sense, although we'll need to refactor the whole codebase for it. #399

File: session/pingpong/invoice_tracker.go (hunk 11 above)
Comment (vkuznecovas) [reply session/pingpong/invoice_tracker.go]: https://github.com/mysteriumnetwork/payments/issues/34

File: session/pingpong/factory.go (hunk 10 above)
Comment (vkuznecovas) [reply session/pingpong/factory.go]: I'll add in docs in a separate PR documenting why this is the case.

**Pull request merged**